In [4]:
import pandas as pd
import os
import re

path = "resources/Tablas_elecciones_barrio/"
files = ""
tables = []


In [5]:
# Cargar los nombre de todos los archivos almacenados dentro de la carpeta "resources/Tablas_elecciones_barrio/"
for r, d, f in os.walk(path):
    files = f
# lectura y limpieza de las 21 tablas. Resultado: una lista de 21 tablas
for file in files:
    tmp_table = pd.read_excel(path + '{0}'.format(file), 
                              sep=";", encoding="ISO-8859-1")
    tmp_table.columns = [i.strip().replace(' ','_') for i in tmp_table.columns]
    tmp_table['Distrito'] = re.sub(" ", "", re.sub("^[0-9]+(\.)(\s)", "", tmp_table.Barrio[0].upper()))
#     print(re.sub(" ", "",re.sub("^[0-9]+(\.)(\s)", "", tmp_table.Barrio[0].upper())))
    tmp_table.drop([0,1], inplace=True) 
    for i in range(0, len(tmp_table.Barrio)):
        tmp_table.iloc[i,0] = re.sub("[0-9]+(\.)(\s)", "", str.strip(tmp_table.iloc[i,0].upper()))
    
    tables.append(tmp_table)

# Se concatenan las 21 tablas en una sola
main_table_elecciones = pd.concat(tables, sort=False)

# Se unifica el index para toda la tabla
main_table_elecciones.reset_index(drop=True, inplace=True)    
main_table_elecciones.head(10)


Barrio  Censo_electoral  Abstención  Votos_nulos  \
0         CASA DE CAMPO           10.648       2.726         26.0   
1             ARGÜELLES           19.127       5.053         36.0   
2  CIUDAD UNIVERSITARIA           12.669       2.890         32.0   
3            VALDEZARZA           23.044       6.735         68.0   
4            VALDEMARÍN            4.267     909.000          5.0   
5            EL PLANTÍO            1.989     465.000          7.0   
6               ARAVACA           18.851       4.032         30.0   
7            GAZTAMBIDE           18.447       4.910         35.0   
8              ARAPILES           19.472       5.163         52.0   
9             TRAFALGAR           19.322       5.359         35.0   

   Votos_blancos  Total_votos_a_candidaturas  PACMA   PH  P.S.O.E.     P.P.  \
0           33.0                       7.863   34.0  2.0   718.000    1.910   
1           59.0                      13.979   40.0  4.0   956.000    4.882   
2           31.0                       9.716   30.0  7.0   680.000    3.580   
3           81.0                      16.160   98.0  5.0     2.379    3.577   
4            3.0                       3.350    7.0  0.0    93.000    1.773   
5            3.0                       1.514    1.0  1.0    65.000  735.000   
6           64.0                      14.725   33.0  2.0     1.015    5.645   
7           60.0                      13.442   46.0  3.0     1.049    4.414   
8           66.0                      14.191   57.0  3.0     1.255    4.132   
9           57.0                      13.871   48.0  2.0     1.184    3.821   

   ...  IU-MpM  MCE      VOX  UPyD  M.I.  P.C.P.E.  CNTG+  PCTE  \
0  ...   181.0  1.0  541.000  14.0   0.0       3.0   10.0   4.0   
1  ...   214.0  4.0    1.394  15.0   6.0       1.0   10.0   0.0   
2  ...   166.0  3.0  843.000   8.0   0.0       1.0   11.0   2.0   
3  ...   436.0  2.0    1.032  18.0   6.0       6.0   21.0   3.0   
4  ...    13.0  0.0  466.000   0.0   1.0       0.0    6.0   0.0   
5  ...    10.0  0.0  227.000   0.0   0.0       0.0    0.0   0.0   
6  ...   241.0  1.0    1.443  11.0   1.0       3.0    7.0   3.0   
7  ...   243.0  6.0    1.209  17.0   0.0       5.0   18.0   3.0   
8  ...   273.0  5.0    1.093  17.0   0.0       5.0    9.0   2.0   
9  ...   288.0  2.0    1.039  12.0   2.0       3.0    8.0   3.0   

   PCAS-TC-PPCCAL-PACTO         Distrito  
0                   2.0  MONCLOA-ARAVACA  
1                   6.0  MONCLOA-ARAVACA  
2                   2.0  MONCLOA-ARAVACA  
3                  14.0  MONCLOA-ARAVACA  
4                   0.0  MONCLOA-ARAVACA  
5                   0.0  MONCLOA-ARAVACA  
6                   1.0  MONCLOA-ARAVACA  
7                   5.0         CHAMBERI  
8                   7.0         CHAMBERI  
9                  11.0         CHAMBERI  

[10 rows x 26 columns]

In [6]:
import pandas as pd

# Cargar la tabla del Padron por barrios
padron = pd.read_excel("resources/padron_barrios.xls")


In [7]:
# Limpieza y organización de la tabla
tmp = []
separador_tabla = 0
for i in range(0, len(padron.Españoles_Hombres)):
    if pd.isna(padron.Españoles_Hombres[i]):
        dis_bar = padron.Edad[i].replace(" ", "").split("/")
        for j in  range(i+1, len(padron.Españoles_Hombres)):
            if pd.isna(padron.Españoles_Hombres[j]):
                tmp.append(padron.iloc[i:j,].copy(deep=True))
                tmp[separador_tabla].drop([i,i+1], inplace=True) 
                tmp[separador_tabla]['Distrito'] = dis_bar[0]
                tmp[separador_tabla]['Barrio'] = dis_bar[1].strip()
                break
        separador_tabla += 1

main_table_padron = pd.concat(tmp)
main_table_padron.Barrio = [str.strip(i) for i in main_table_padron.Barrio]
main_table_padron.reset_index(drop=True, inplace=True)
main_table_padron.head(50)

Edad  Españoles_Hombres  Españoles_Mujeres  Españoles_Total  \
0       0 - 4              263.0              238.0            501.0   
1       5 - 9              235.0              265.0            500.0   
2     10 - 14              227.0              238.0            465.0   
3     15 - 19              234.0              282.0            516.0   
4     20 - 24              323.0              346.0            669.0   
5     25 - 29              627.0              567.0           1194.0   
6     30 - 34              833.0              729.0           1562.0   
7     35 - 39              850.0              759.0           1609.0   
8     40 - 44             1002.0              826.0           1828.0   
9     45 - 49              931.0              772.0           1703.0   
10    50 - 54              825.0              760.0           1585.0   
11    55 - 59              775.0              751.0           1526.0   
12    60 - 64              663.0              740.0           1403.0   
13    65 - 69              507.0              608.0           1115.0   
14    70 - 74              420.0              507.0            927.0   
15    75 - 79              338.0              463.0            801.0   
16    80 - 84              177.0              372.0            549.0   
17    85 - 89              143.0              346.0            489.0   
18    90 - 94               64.0              194.0            258.0   
19    95 - 99               16.0               59.0             75.0   
20  100 o más                2.0               17.0             19.0   
21      0 - 4              442.0              475.0            917.0   
22      5 - 9              491.0              439.0            930.0   
23    10 - 14              468.0              456.0            924.0   
24    15 - 19              465.0              410.0            875.0   
25    20 - 24              561.0              539.0           1100.0   
26    25 - 29             1195.0             1299.0           2494.0   
27    30 - 34             1606.0             1569.0           3175.0   
28    35 - 39             1873.0             1617.0           3490.0   
29    40 - 44             1832.0             1608.0           3440.0   
30    45 - 49             1712.0             1414.0           3126.0   
31    50 - 54             1492.0             1226.0           2718.0   
32    55 - 59             1250.0             1262.0           2512.0   
33    60 - 64              996.0             1108.0           2104.0   
34    65 - 69              673.0              815.0           1488.0   
35    70 - 74              553.0              794.0           1347.0   
36    75 - 79              429.0              652.0           1081.0   
37    80 - 84              312.0              528.0            840.0   
38    85 - 89              234.0              599.0            833.0   
39    90 - 94               94.0              305.0            399.0   
40    95 - 99               24.0               97.0            121.0   
41  100 o más                1.0               18.0             19.0   
42      0 - 4              111.0              112.0            223.0   
43      5 - 9               92.0               85.0            177.0   
44    10 - 14               99.0               80.0            179.0   
45    15 - 19              115.0               88.0            203.0   
46    20 - 24              137.0              141.0            278.0   
47    25 - 29              289.0              272.0            561.0   
48    30 - 34              368.0              353.0            721.0   
49    35 - 39              409.0              338.0            747.0   

    Extranjeros_Hombres  Extranjeros_Mujeres  Extranjeros_Total  \
0                  56.0                 50.0              106.0   
1                  43.0                 38.0               81.0   
2                  30.0                 38.0               68.0   
3                  44.0                 55.0               99.0   
4

## Resultados electorales por porcentaje de migrantes

### Votantes Vox

In [53]:
# Combinación de columnas
datos_unidos = main_table_elecciones.merge(main_table_padron, on=["Distrito", "Barrio"])

# Elimino los que no tienen edad de votar
datos_votantes = datos_unidos[(datos_unidos.Edad != '0 - 4')
                              & (datos_unidos.Edad != '5 - 9')
                              & (datos_unidos.Edad != '10 - 14')
                              & (datos_unidos.Edad != '15 - 19')]
datos_votantes

Barrio  Censo_electoral  Abstención  Votos_nulos  Votos_blancos  \
4     VALDEZARZA           23.044       6.735         68.0           81.0   
5     VALDEZARZA           23.044       6.735         68.0           81.0   
6     VALDEZARZA           23.044       6.735         68.0           81.0   
7     VALDEZARZA           23.044       6.735         68.0           81.0   
8     VALDEZARZA           23.044       6.735         68.0           81.0   
...          ...              ...         ...          ...            ...   
1568      ATOCHA          840.000     217.000          0.0            0.0   
1569      ATOCHA          840.000     217.000          0.0            0.0   
1570      ATOCHA          840.000     217.000          0.0            0.0   
1571      ATOCHA          840.000     217.000          0.0            0.0   
1572      ATOCHA          840.000     217.000          0.0            0.0   

      Total_votos_a_candidaturas  PACMA   PH  P.S.O.E.    P.P.  ...  \
4                          16.16   98.0  5.0     2.379   3.577  ...   
5                          16.16   98.0  5.0     2.379   3.577  ...   
6                          16.16   98.0  5.0     2.379   3.577  ...   
7                          16.16   98.0  5.0     2.379   3.577  ...   
8                          16.16   98.0  5.0     2.379   3.577  ...   
...                          ...    ...  ...       ...     ...  ...   
1568                      623.00    0.0  1.0    85.000  73.000  ...   
1569                      623.00    0.0  1.0    85.000  73.000  ...   
1570                      623.00    0.0  1.0    85.000  73.000  ...   
1571                      623.00    0.0  1.0    85.000  73.000  ...   
1572                      623.00    0.0  1.0    85.000  73.000  ...   

           Edad  Españoles_Hombres  Españoles_Mujeres  Españoles_Total  \
4       20 - 24              552.0              540.0           1092.0   
5       25 - 29              652.0              717.0           1369.0   
6       30 - 34              771.0              738.0           1509.0   
7       35 - 39              858.0              853.0           1711.0   
8       40 - 44              923.0              989.0           1912.0   
...         ...                ...                ...              ...   
1568    80 - 84                6.0               12.0             18.0   
1569    85 - 89                1.0                6.0              7.0   
1570    90 - 94                1.0                1.0              2.0   
1571    95 - 99                0.0                1.0              1.0   
1572  100 o más                1.0                0.0              1.0   

      Extranjeros_Hombres  Extranjeros_Mujeres  Extranjeros_Total  \
4                   179.0                249.0              428.0   
5                   200.0                306.0              506.0   
6                   192.0                280.0              472.0   
7                   208.0                227.0              435.0   
8                   163.0                195.0              358.0   
...                   ...                  ...                ...   
1568                  0.0                  0.0                0.0   
1569                  0.0                  0.0                0.0   
1570                  0.0                  0.0                0.0   
1571                  0.0                  0.0                0.0   
1572                  0.0                  0.0                0.0   

      Total_Hombres  Total_Mujeres   Total  
4             731.0          789.0  1520.0  
5             852.0         1023.0  1875.0  
6             963.0         1018.0  1981.0  
7            1066.0         1080.0  2146.0  
8            1086.0         1184.0  2270.0  
...             ...            ...     ...  
1568            6.0           12.0    18.0  
1569            1.0            6.0     7.0  
1570            1.0            1.0     2.0  
1571            0.0            1.0     1.0  
1572            1.0

In [68]:
# Extraigo votantes de VOX
datos_vox = datos_votantes[['Distrito', 'Censo_electoral', 'Abstención', 'Votos_nulos', 'Votos_blancos','VOX','Españoles_Total', 'Extranjeros_Total','Total']]

# Agrupo por distritos y calculo el porcentaje de extranjeros y votantes VOX
datos_vox2 = datos_vox.groupby('Distrito').agg(
    {'VOX': 'max', 'Censo_electoral': 'max', 'Españoles_Total': 'sum', 'Extranjeros_Total': 'sum', 'Total': 'sum'}).reset_index()

datos_vox3 = datos_vox2.assign(
    porcentaje_extranjeros = datos_vox2.Total / datos_vox2.Extranjeros_Total, 
    porcentaje_vox = datos_vox2.Censo_electoral / datos_vox2.VOX)

datos_vox3

datos_vox3.sort_values(['porcentaje_extranjeros'], ascending=False)


Distrito      VOX  Censo_electoral  Españoles_Total  \
18           VICALVARO  556.000           13.799          13188.0   
13              RETIRO  868.000           19.263          49533.0   
11           MORATALAZ  517.000           22.729          46392.0   
0           ARGANZUELA  739.000          840.000          97965.0   
4            CHAMARTIN    1.891           29.231          65340.0   
7   FUENCARRAL-ELPARDO  215.000           41.495         100551.0   
8            HORTALEZA  760.000           37.395          83393.0   
10     MONCLOA-ARAVACA    1.443           23.044          40553.0   
5             CHAMBERI    1.365           19.472          86096.0   
15  SANBLAS-CANILLEJAS  965.000           23.927         102838.0   
14           SALAMANCA    2.284           32.881          92388.0   
6         CIUDADLINEAL  810.000           35.549          74212.0   
19          VILLAVERDE  547.000           11.910          11929.0   
9               LATINA    2.259           49.544          86406.0   
2          CARABANCHEL  949.000           31.833          88736.0   
12    PUENTEDEVALLECAS  681.000           32.438          51583.0   
16              TETUAN  869.000           17.472          65005.0   
1              BARAJAS   70.000            1.244           1199.0   
17               USERA  618.000           17.019          56468.0   
3               CENTRO  949.000           31.496          95806.0   

    Extranjeros_Total     Total  porcentaje_extranjeros  porcentaje_vox  
18             1062.0   14250.0               13.418079        0.024818  
13             4496.0   54029.0               12.017126        0.022192  
11             4577.0   50969.0               11.135897        0.043963  
0             10752.0  108717.0               10.111328        1.136671  
4              7229.0   72569.0               10.038595       15.457959  
7             11389.0  111940.0                9.828782        0.193000  
8             10321.0   93714.0                9.079934        0.049204  
10             5323.0   45876.0                8.618448       15.969508  
5             12233.0   98329.0                8.038012       14.265201  
15            17193.0  120031.0                6.981388        0.024795  
14            16416.0  108804.0                6.627924       14.396235  
6             13431.0   87643.0                6.525426        0.043888  
19             2314.0   14243.0                6.155143        0.021773  
9             18762.0  105168.0                5.605373       21.931828  
2             21365.0  110101.0                5.153335        0.033544  
12            12790.0   64373.0                5.033073        0.047633  
16            17004.0   82009.0                4.822924        0.020106  
1               366.0    1565.0                4.275956        0.017771  
17            17269.0   73737.0                4.269906        0.027539  
3             30814.0  126620.0                4.109171        0.033189

No podemos concluir que el porcentaje de población extranjera por distrito es proporcional a los votos que recibe el partido VOX. 
Siendo **Vicálvaro** el distrito con más extranjeros, dicho partido político tiene únicamente un 2.48%. Mientras que los distritos donde el peso de VOX es mayor, como Latina, Salamanca, Chamberí, Chamartín o Moncloa-Aravaca no se encuentran en entre los 3 barrios con menos españoles.

In [81]:
# Folium para mapas
import folium

# Mapa base sobre Madrid
map_madrid = folium.Map(location=[40.42, -3.7], zoom_start=13)

# COROPLETAS
folium.Choropleth(
    geo_data='resources/neighbourhoods.geojson',
    data=datos_vox3,
    columns=['Distrito', 'VOX'],
    key_on='feature.properties.neighbourhood',
    fill_color='YlGn',
    bins=[0, 200, 400, 600, 800, 1000],
).add_to(map_madrid)

map_madrid

Estudiemos ahora, dentro de dichos distritos, los distintos barrios. De este modo, pretendemos demostrar que aumentan los votos a VOX en las zonas limítrofes a áreas de migrantes.

### LATINA POR BARRIOS

In [102]:
# Extraigo votantes de Barrios de Latina
datos_latina = datos_votantes[datos_votantes.Distrito == 'LATINA']

datos_latina2 = datos_latina[['Barrio', 'Censo_electoral', 'Abstención', 'Votos_nulos', 'Votos_blancos','VOX','Españoles_Total', 'Extranjeros_Total','Total']]

# Agrupo por distritos y calculo el porcentaje de extranjeros y votantes VOX
datos_latina3 = datos_latina2.groupby('Barrio').agg(
    {'VOX': 'max', 'Censo_electoral': 'max', 'Españoles_Total': 'sum', 'Extranjeros_Total': 'sum', 'Total': 'sum'}).reset_index()

datos_latina4 = datos_latina3.assign(
    porcentaje_extranjeros = datos_latina3.Total / datos_latina3.Extranjeros_Total, 
    porcentaje_vox = datos_latina3.Censo_electoral / datos_latina3.VOX)

datos_latina4


Barrio    VOX  Censo_electoral  Españoles_Total  Extranjeros_Total  \
0      ALUCHE  2.259           49.544          48042.0             9634.0   
1  CAMPAMENTO  1.028           13.646          13184.0             3462.0   
2      LUCERO  1.023           26.056          25180.0             5666.0   

     Total  porcentaje_extranjeros  porcentaje_vox  
0  57676.0                5.986714       21.931828  
1  16646.0                4.808203       13.274319  
2  30846.0                5.444052       25.470186

### MONCLOA-ARAVACA POR BARRIOS

### CHAMARTÍN POR BARRIOS

In [63]:
datos_vox3.dtypes

VOX                       float64
Censo_electoral           float64
Españoles_Total           float64
Extranjeros_Total         float64
Total                     float64
porcentaje_extranjeros    float64
porcentaje_vox            float64
dtype: object

In [64]:
import folium

# Incializamos el mapa
mapa_mdd = folium.Map(location=[40.42, -3.7], zoom_start=13, tiles='cartodbpositron')

mapa_mdd

In [ ]:
import numpy as np

In [ ]:
import seaborn as sns

In [84]:
# Folium para mapas
import folium

# Mapa base sobre Madrid
map_madd = folium.Map(location=[40.42, -3.7], zoom_start=13)

# Heatmap para mapa de calor
from folium.plugins import HeatMap

#heatmap = HeatMap(data=datos_vox3[['Distrito', 'VOX']], radius=15).add_to(map_madd)
#map_madd

8. Hay alguna relación entre el numero de migrantes y PSOE/VOX 

3. Rdos por %migrantes y distrito : Inma